File: functional_programming_triller/exercicio_aula_2/poo/Cliente.java
package poo;
import java.util.ArrayList;

public class Cliente extends Pessoa implements Comprador {
  private final int LIMITE_SERASA = 800;
  private int scoreSerasa;
  private ArrayList<Produto> produtos;

  public Cliente(String nome, String sobrenome, String pais, int anoNascimento, int score) {
    this.nome = nome;
    this.sobrenome = sobrenome;
    this.anoNascimento = anoNascimento;
    this.pais = pais;
    this.scoreSerasa = score;
    this.produtos = new ArrayList<>();
  }

  @Override
  public int quantidadeComprada() {
    return this.produtos.size();
  }

  @Override
  public void comprar(Produto produto) {
    if(podeComprar()) {
      this.produtos.add(produto);
    } else {
      System.out.println("Esse cliente está com serasa zuado!");
    }
  }

  @Override
  public void produtosComprados() {
    for(Produto produto : produtos) {
      System.out.println("Nome: " + produto.titulo);
      System.out.println("Preço Pago: " + (produto.preco + produto.calculaImposto()));
    }
  }

  @Override
  public boolean podeComprar() {
    if(this.scoreSerasa < LIMITE_SERASA) {
      return false;
    }

    return true;
  }
}

File: functional_programming_triller/exercicio_aula_2/functional/cliente.exs
defmodule Cliente do
  @limit_serasa 800

  defstruct [
    nome: "",
    sobrenome: "",
    pais: "",
    ano_nascimento: 0,
    score_serasa: 0
  ]

  def quantidade_comprada(produtos) do
    Enum.count(produtos)
  end

  def comprar(%Cliente{} = cliente, produtos_comprados, produto) do
    if pode_comprar? do
      produtos_comprados ++ produto
    else
      IO.puts("Esse cliente possui score baixo no serasa")
    end
  end

  def produtos_comprados(%Suplemento{} = suplemento) do
    IO.puts("Nome: ", suplemento.titulo)
    IO.puts("Preço: ", (suplemento.preco + Suplemento.calcula_imposto(suplemento.preco)))
  end

  def produtos_comprados(produtos) when is_list(produtos) do
    [_ | produtos_restantes] = produtos
    produtos_comprados(produtos_restantes)
  end

  defp pode_comprar?(cliente) do
    if cliente.score_serasa < @limit_serasa do
      true
    else
      false
  end
end
